##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# 使用 Keras 和 Tensorflow Hub 对电影评论进行文本分类

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://tensorflow.google.cn/tutorials/keras/text_classification_with_hub"><img src="https://tensorflow.google.cn/images/tf_logo_32px.png" />在 tensorFlow.google.cn 上查看</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/colab_logo_32px.png" />在 Google Colab 中运行</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/GitHub-Mark-32px.png" />在 GitHub 上查看源代码</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/download_logo_32px.png" />下载 notebook</a>
  </td>
</table>

Note: 我们的 TensorFlow 社区翻译了这些文档。因为社区翻译是尽力而为， 所以无法保证它们是最准确的，并且反映了最新的
[官方英文文档](https://tensorflow.google.cn/?hl=en)。如果您有改进此翻译的建议， 请提交 pull request 到
[tensorflow/docs](https://github.com/tensorflow/docs) GitHub 仓库。要志愿地撰写或者审核译文，请加入
[docs-zh-cn@tensorflow.org Google Group](https://groups.google.com/a/tensorflow.org/forum/#!forum/docs-zh-cn)。

此笔记本（notebook）使用评论文本将影评分为*积极（positive）*或*消极（nagetive）*两类。这是一个*二元（binary）*或者二分类问题，一种重要且应用广泛的机器学习问题。

本教程演示了使用 Tensorflow Hub 和 Keras 进行迁移学习的基本应用。

我们将使用来源于[网络电影数据库（Internet Movie Database）](https://www.imdb.com/)的 [IMDB 数据集（IMDB dataset）](https://tensorflow.google.cn/api_docs/python/tf/keras/datasets/imdb)，其包含 50,000 条影评文本。从该数据集切割出的 25,000 条评论用作训练，另外 25,000 条用作测试。训练集与测试集是*平衡的（balanced）*，意味着它们包含相等数量的积极和消极评论。

此笔记本（notebook）使用了 [tf.keras](https://tensorflow.google.cn/guide/keras)，它是一个 Tensorflow 中用于构建和训练模型的高级API，此外还使用了 [TensorFlow Hub](https://tensorflow.google.cn/hub)，一个用于迁移学习的库和平台。有关使用 `tf.keras` 进行文本分类的更高级教程，请参阅 [MLCC文本分类指南（MLCC Text Classification Guide）](https://developers.google.com/machine-learning/guides/text-classification/)。

In [3]:
import numpy as np

import tensorflow as tf

!pip install -q tensorflow-hub
!pip install -q tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

You should consider upgrading via the '/tmpfs/src/tf_docs_env/bin/python -m pip install --upgrade pip' command.


You should consider upgrading via the '/tmpfs/src/tf_docs_env/bin/python -m pip install --upgrade pip' command.


Version:  2.3.0
Eager mode:  True
Hub version:  0.9.0


GPU is available


## 下载 IMDB 数据集
IMDB数据集可以在 [Tensorflow 数据集](https://github.com/tensorflow/datasets)处获取。以下代码将 IMDB 数据集下载至您的机器（或 colab 运行时环境）中：

In [4]:
# 将训练集分割成 60% 和 40%，从而最终我们将得到 15,000 个训练样本
# 10,000 个验证样本以及 25,000 个测试样本。
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteZDZ3AR/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteZDZ3AR/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incompleteZDZ3AR/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## 探索数据

让我们花一点时间来了解数据的格式。每一个样本都是一个表示电影评论和相应标签的句子。该句子不以任何方式进行预处理。标签是一个值为 0 或 1 的整数，其中 0 代表消极评论，1 代表积极评论。

我们来打印下前十个样本。

In [5]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

我们再打印下前十个标签。

In [6]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## 构建模型

神经网络由堆叠的层来构建，这需要从三个主要方面来进行体系结构决策：

* 如何表示文本？
* 模型里有多少层？
* 每个层里有多少*隐层单元（hidden units）*？

本示例中，输入数据由句子组成。预测的标签为 0 或 1。

表示文本的一种方式是将句子转换为嵌入向量（embeddings vectors）。我们可以使用一个预先训练好的文本嵌入（text embedding）作为首层，这将具有三个优点：

* 我们不必担心文本预处理
* 我们可以从迁移学习中受益
* 嵌入具有固定长度，更易于处理

针对此示例我们将使用 [TensorFlow Hub](https://tensorflow.google.cn/hub) 中名为 [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1) 的一种**预训练文本嵌入（text embedding）模型** 。

为了达到本教程的目的还有其他三种预训练模型可供测试：

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) ——类似 [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)，但 2.5%的词汇转换为未登录词桶（OOV buckets）。如果任务的词汇与模型的词汇没有完全重叠，这将会有所帮助。
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) ——一个拥有约 1M 词汇量且维度为 50 的更大的模型。
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) ——拥有约 1M 词汇量且维度为128的更大的模型。

让我们首先创建一个使用 Tensorflow Hub 模型嵌入（embed）语句的Keras层，并在几个输入样本中进行尝试。请注意无论输入文本的长度如何，嵌入（embeddings）输出的形状都是：`(num_examples, embedding_dimension)`。


In [7]:
embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 20), dtype=float32, numpy=
array([[ 1.765786  , -3.882232  ,  3.9134233 , -1.5557289 , -3.3362343 ,
        -1.7357955 , -1.9954445 ,  1.2989551 ,  5.081598  , -1.1041286 ,
        -2.0503852 , -0.72675157, -0.65675956,  0.24436149, -3.7208383 ,
         2.0954835 ,  2.2969332 , -2.0689783 , -2.9489717 , -1.1315987 ],
       [ 1.8804485 , -2.5852382 ,  3.4066997 ,  1.0982676 , -4.056685  ,
        -4.891284  , -2.785554  ,  1.3874227 ,  3.8476458 , -0.9256538 ,
        -1.896706  ,  1.2113281 ,  0.11474707,  0.76209456, -4.8791065 ,
         2.906149  ,  4.7087674 , -2.3652055 , -3.5015898 , -1.6390051 ],
       [ 0.71152234, -0.6353217 ,  1.7385626 , -1.1168286 , -0.5451594 ,
        -1.1808156 ,  0.09504455,  1.4653089 ,  0.66059524,  0.79308075,
        -2.2268345 ,  0.07446612, -1.4075904 , -0.70645386, -1.907037  ,
         1.4419787 ,  1.9551861 , -0.42660055, -2.8022065 ,  0.43727064]],
      dtype=float32)>

现在让我们构建完整模型：

In [8]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


层按顺序堆叠以构建分类器：

1. 第一层是 Tensorflow Hub 层。这一层使用一个预训练的保存好的模型来将句子映射为嵌入向量（embedding vector）。我们所使用的预训练文本嵌入（embedding）模型([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1))将句子切割为符号，嵌入（embed）每个符号然后进行合并。最终得到的维度是：`(num_examples, embedding_dimension)`。
2. 该定长输出向量通过一个有 16 个隐层单元的全连接层（`Dense`）进行管道传输。
3. 最后一层与单个输出结点紧密相连。使用 `Sigmoid` 激活函数，其函数值为介于 0 与 1 之间的浮点数，表示概率或置信水平。

让我们编译模型。

### 损失函数与优化器

一个模型需要损失函数和优化器来进行训练。由于这是一个二分类问题且模型输出概率值（一个使用 sigmoid 激活函数的单一单元层），我们将使用 `binary_crossentropy` 损失函数。

这不是损失函数的唯一选择，例如，您可以选择 `mean_squared_error` 。但是，一般来说 `binary_crossentropy` 更适合处理概率——它能够度量概率分布之间的“距离”，或者在我们的示例中，指的是度量 ground-truth 分布与预测值之间的“距离”。

稍后，当我们研究回归问题（例如，预测房价）时，我们将介绍如何使用另一种叫做均方误差的损失函数。

现在，配置模型来使用优化器和损失函数：

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## 训练模型

以 512 个样本的 mini-batch 大小迭代 20 个 epoch 来训练模型。 这是指对 `x_train` 和 `y_train` 张量中所有样本的的 20 次迭代。在训练过程中，监测来自验证集的 10,000 个样本上的损失值（loss）和准确率（accuracy）：

In [10]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=20,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/20


 1/30 [>.............................] - ETA: 0s - loss: 2.3139 - accuracy: 0.5176

 2/30 [=>............................] - ETA: 0s - loss: 2.2724 - accuracy: 0.5117

 3/30 [==>...........................] - ETA: 0s - loss: 2.1627 - accuracy: 0.5280

 5/30 [====>.........................] - ETA: 1s - loss: 2.1952 - accuracy: 0.5039

 7/30 [======>.......................] - ETA: 0s - loss: 2.1435 - accuracy: 0.4986

 9/30 [========>.....................] - ETA: 0s - loss: 2.0707 - accuracy: 0.5011

11/30 [==========>...................] - ETA: 0s - loss: 2.0046 - accuracy: 0.5016

13/30 [============>.................] - ETA: 0s - loss: 1.9295 - accuracy: 0.5056

15/30 [==============>...............] - ETA: 0s - loss: 1.8852 - accuracy: 0.5026

17/30 [================>.............] - ETA: 0s - loss: 1.8295 - accuracy: 0.5020

19/30 [==================>...........] - ETA: 0s - loss: 1.7831 - accuracy: 0.4978

21/30 [====================>.........] - ETA: 0s - loss: 1.7286 - accuracy: 0.4998

23/30 [======================>.......] - ETA: 0s - loss: 1.6797 - accuracy: 0.5003

25/30 [========================>.....] - ETA: 0s - loss: 1.6320 - accuracy: 0.4991

27/30 [==========================>...] - ETA: 0s - loss: 1.5894 - accuracy: 0.4983

29/30 [============================>.] - ETA: 0s - loss: 1.5505 - accuracy: 0.4962

30/30 [==============================] - 2s 64ms/step - loss: 1.5444 - accuracy: 0.4965 - val_loss: 0.9259 - val_accuracy: 0.4705


Epoch 2/20


 1/30 [>.............................] - ETA: 0s - loss: 0.9305 - accuracy: 0.4727

 3/30 [==>...........................] - ETA: 0s - loss: 0.9075 - accuracy: 0.4655

 5/30 [====>.........................] - ETA: 0s - loss: 0.8813 - accuracy: 0.4699

 7/30 [======>.......................] - ETA: 0s - loss: 0.8654 - accuracy: 0.4674

 9/30 [========>.....................] - ETA: 0s - loss: 0.8497 - accuracy: 0.4701

11/30 [==========>...................] - ETA: 0s - loss: 0.8383 - accuracy: 0.4709

13/30 [============>.................] - ETA: 0s - loss: 0.8263 - accuracy: 0.4707

15/30 [==============>...............] - ETA: 0s - loss: 0.8145 - accuracy: 0.4740

17/30 [================>.............] - ETA: 0s - loss: 0.8039 - accuracy: 0.4762

19/30 [==================>...........] - ETA: 0s - loss: 0.7956 - accuracy: 0.4791

21/30 [====================>.........] - ETA: 0s - loss: 0.7892 - accuracy: 0.4860

23/30 [======================>.......] - ETA: 0s - loss: 0.7837 - accuracy: 0.4890

25/30 [========================>.....] - ETA: 0s - loss: 0.7771 - accuracy: 0.4938

27/30 [==========================>...] - ETA: 0s - loss: 0.7717 - accuracy: 0.4958

29/30 [============================>.] - ETA: 0s - loss: 0.7668 - accuracy: 0.4989

30/30 [==============================] - 2s 59ms/step - loss: 0.7667 - accuracy: 0.4990 - val_loss: 0.7017 - val_accuracy: 0.5327


Epoch 3/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6893 - accuracy: 0.5312

 3/30 [==>...........................] - ETA: 0s - loss: 0.6868 - accuracy: 0.5475

 5/30 [====>.........................] - ETA: 0s - loss: 0.6884 - accuracy: 0.5504

 7/30 [======>.......................] - ETA: 0s - loss: 0.6837 - accuracy: 0.5564

 9/30 [========>.....................] - ETA: 0s - loss: 0.6784 - accuracy: 0.5579

11/30 [==========>...................] - ETA: 0s - loss: 0.6795 - accuracy: 0.5513

13/30 [============>.................] - ETA: 0s - loss: 0.6797 - accuracy: 0.5527

15/30 [==============>...............] - ETA: 0s - loss: 0.6780 - accuracy: 0.5538

17/30 [================>.............] - ETA: 0s - loss: 0.6743 - accuracy: 0.5560

19/30 [==================>...........] - ETA: 0s - loss: 0.6725 - accuracy: 0.5591

21/30 [====================>.........] - ETA: 0s - loss: 0.6713 - accuracy: 0.5636

23/30 [======================>.......] - ETA: 0s - loss: 0.6700 - accuracy: 0.5654

25/30 [========================>.....] - ETA: 0s - loss: 0.6662 - accuracy: 0.5719

27/30 [==========================>...] - ETA: 0s - loss: 0.6645 - accuracy: 0.5739

29/30 [============================>.] - ETA: 0s - loss: 0.6632 - accuracy: 0.5790

30/30 [==============================] - 2s 58ms/step - loss: 0.6631 - accuracy: 0.5799 - val_loss: 0.6387 - val_accuracy: 0.6238


Epoch 4/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6175 - accuracy: 0.6387

 3/30 [==>...........................] - ETA: 0s - loss: 0.6235 - accuracy: 0.6393

 5/30 [====>.........................] - ETA: 0s - loss: 0.6285 - accuracy: 0.6254

 7/30 [======>.......................] - ETA: 0s - loss: 0.6268 - accuracy: 0.6331

 9/30 [========>.....................] - ETA: 0s - loss: 0.6235 - accuracy: 0.6289

11/30 [==========>...................] - ETA: 0s - loss: 0.6283 - accuracy: 0.6200

13/30 [============>.................] - ETA: 0s - loss: 0.6244 - accuracy: 0.6235

15/30 [==============>...............] - ETA: 0s - loss: 0.6226 - accuracy: 0.6242

17/30 [================>.............] - ETA: 0s - loss: 0.6213 - accuracy: 0.6273

19/30 [==================>...........] - ETA: 0s - loss: 0.6218 - accuracy: 0.6283

21/30 [====================>.........] - ETA: 0s - loss: 0.6205 - accuracy: 0.6301

23/30 [======================>.......] - ETA: 0s - loss: 0.6196 - accuracy: 0.6303

25/30 [========================>.....] - ETA: 0s - loss: 0.6182 - accuracy: 0.6320

27/30 [==========================>...] - ETA: 0s - loss: 0.6170 - accuracy: 0.6313

29/30 [============================>.] - ETA: 0s - loss: 0.6154 - accuracy: 0.6326

30/30 [==============================] - 2s 58ms/step - loss: 0.6156 - accuracy: 0.6327 - val_loss: 0.6051 - val_accuracy: 0.6390


Epoch 5/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6132 - accuracy: 0.6387

 3/30 [==>...........................] - ETA: 0s - loss: 0.6079 - accuracy: 0.6426

 5/30 [====>.........................] - ETA: 0s - loss: 0.5998 - accuracy: 0.6539

 7/30 [======>.......................] - ETA: 0s - loss: 0.5937 - accuracy: 0.6490

 9/30 [========>.....................] - ETA: 0s - loss: 0.5922 - accuracy: 0.6480

11/30 [==========>...................] - ETA: 0s - loss: 0.5920 - accuracy: 0.6500

13/30 [============>.................] - ETA: 0s - loss: 0.5907 - accuracy: 0.6532

15/30 [==============>...............] - ETA: 0s - loss: 0.5895 - accuracy: 0.6543

17/30 [================>.............] - ETA: 0s - loss: 0.5853 - accuracy: 0.6597

19/30 [==================>...........] - ETA: 0s - loss: 0.5853 - accuracy: 0.6616

21/30 [====================>.........] - ETA: 0s - loss: 0.5848 - accuracy: 0.6610

23/30 [======================>.......] - ETA: 0s - loss: 0.5849 - accuracy: 0.6624

25/30 [========================>.....] - ETA: 0s - loss: 0.5847 - accuracy: 0.6619

27/30 [==========================>...] - ETA: 0s - loss: 0.5842 - accuracy: 0.6619

29/30 [============================>.] - ETA: 0s - loss: 0.5820 - accuracy: 0.6627

30/30 [==============================] - 2s 57ms/step - loss: 0.5819 - accuracy: 0.6623 - val_loss: 0.5761 - val_accuracy: 0.6639


Epoch 6/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5728 - accuracy: 0.6777

 3/30 [==>...........................] - ETA: 0s - loss: 0.5679 - accuracy: 0.6732

 5/30 [====>.........................] - ETA: 0s - loss: 0.5629 - accuracy: 0.6863

 7/30 [======>.......................] - ETA: 0s - loss: 0.5615 - accuracy: 0.6897

 9/30 [========>.....................] - ETA: 0s - loss: 0.5536 - accuracy: 0.6986

11/30 [==========>...................] - ETA: 0s - loss: 0.5536 - accuracy: 0.7013

13/30 [============>.................] - ETA: 0s - loss: 0.5521 - accuracy: 0.7010

15/30 [==============>...............] - ETA: 0s - loss: 0.5531 - accuracy: 0.7023

17/30 [================>.............] - ETA: 0s - loss: 0.5531 - accuracy: 0.7009

19/30 [==================>...........] - ETA: 0s - loss: 0.5520 - accuracy: 0.7022

21/30 [====================>.........] - ETA: 0s - loss: 0.5514 - accuracy: 0.7017

23/30 [======================>.......] - ETA: 0s - loss: 0.5502 - accuracy: 0.7023

25/30 [========================>.....] - ETA: 0s - loss: 0.5492 - accuracy: 0.7020

27/30 [==========================>...] - ETA: 0s - loss: 0.5498 - accuracy: 0.6988

29/30 [============================>.] - ETA: 0s - loss: 0.5493 - accuracy: 0.6985

30/30 [==============================] - 2s 57ms/step - loss: 0.5492 - accuracy: 0.6983 - val_loss: 0.5475 - val_accuracy: 0.6873


Epoch 7/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5374 - accuracy: 0.6992

 3/30 [==>...........................] - ETA: 0s - loss: 0.5281 - accuracy: 0.7083

 5/30 [====>.........................] - ETA: 0s - loss: 0.5328 - accuracy: 0.7109

 7/30 [======>.......................] - ETA: 0s - loss: 0.5340 - accuracy: 0.7076

 9/30 [========>.....................] - ETA: 0s - loss: 0.5311 - accuracy: 0.7159

11/30 [==========>...................] - ETA: 0s - loss: 0.5286 - accuracy: 0.7221

13/30 [============>.................] - ETA: 0s - loss: 0.5250 - accuracy: 0.7261

15/30 [==============>...............] - ETA: 0s - loss: 0.5230 - accuracy: 0.7307

17/30 [================>.............] - ETA: 0s - loss: 0.5219 - accuracy: 0.7332

19/30 [==================>...........] - ETA: 0s - loss: 0.5218 - accuracy: 0.7306

21/30 [====================>.........] - ETA: 0s - loss: 0.5190 - accuracy: 0.7326

23/30 [======================>.......] - ETA: 0s - loss: 0.5177 - accuracy: 0.7317

25/30 [========================>.....] - ETA: 0s - loss: 0.5183 - accuracy: 0.7297

27/30 [==========================>...] - ETA: 0s - loss: 0.5172 - accuracy: 0.7287

29/30 [============================>.] - ETA: 0s - loss: 0.5158 - accuracy: 0.7293

30/30 [==============================] - 2s 58ms/step - loss: 0.5159 - accuracy: 0.7294 - val_loss: 0.5176 - val_accuracy: 0.7277


Epoch 8/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5035 - accuracy: 0.7188

 3/30 [==>...........................] - ETA: 0s - loss: 0.4927 - accuracy: 0.7422

 5/30 [====>.........................] - ETA: 0s - loss: 0.4945 - accuracy: 0.7461

 7/30 [======>.......................] - ETA: 0s - loss: 0.4939 - accuracy: 0.7573

 9/30 [========>.....................] - ETA: 0s - loss: 0.4903 - accuracy: 0.7574

11/30 [==========>...................] - ETA: 0s - loss: 0.4896 - accuracy: 0.7550

13/30 [============>.................] - ETA: 0s - loss: 0.4890 - accuracy: 0.7547

15/30 [==============>...............] - ETA: 0s - loss: 0.4892 - accuracy: 0.7523

17/30 [================>.............] - ETA: 0s - loss: 0.4874 - accuracy: 0.7555

19/30 [==================>...........] - ETA: 0s - loss: 0.4869 - accuracy: 0.7569

21/30 [====================>.........] - ETA: 0s - loss: 0.4873 - accuracy: 0.7570

23/30 [======================>.......] - ETA: 0s - loss: 0.4868 - accuracy: 0.7578

25/30 [========================>.....] - ETA: 0s - loss: 0.4858 - accuracy: 0.7587

27/30 [==========================>...] - ETA: 0s - loss: 0.4830 - accuracy: 0.7603

29/30 [============================>.] - ETA: 0s - loss: 0.4815 - accuracy: 0.7607

30/30 [==============================] - 2s 58ms/step - loss: 0.4813 - accuracy: 0.7609 - val_loss: 0.4884 - val_accuracy: 0.7490


Epoch 9/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4581 - accuracy: 0.8047

 3/30 [==>...........................] - ETA: 0s - loss: 0.4557 - accuracy: 0.7799

 5/30 [====>.........................] - ETA: 0s - loss: 0.4605 - accuracy: 0.7789

 7/30 [======>.......................] - ETA: 0s - loss: 0.4588 - accuracy: 0.7773

 9/30 [========>.....................] - ETA: 0s - loss: 0.4608 - accuracy: 0.7734

11/30 [==========>...................] - ETA: 0s - loss: 0.4609 - accuracy: 0.7706

13/30 [============>.................] - ETA: 0s - loss: 0.4558 - accuracy: 0.7781

15/30 [==============>...............] - ETA: 0s - loss: 0.4568 - accuracy: 0.7798

17/30 [================>.............] - ETA: 0s - loss: 0.4560 - accuracy: 0.7804

19/30 [==================>...........] - ETA: 0s - loss: 0.4544 - accuracy: 0.7827

21/30 [====================>.........] - ETA: 0s - loss: 0.4526 - accuracy: 0.7827

23/30 [======================>.......] - ETA: 0s - loss: 0.4514 - accuracy: 0.7842

25/30 [========================>.....] - ETA: 0s - loss: 0.4503 - accuracy: 0.7837

27/30 [==========================>...] - ETA: 0s - loss: 0.4494 - accuracy: 0.7848

29/30 [============================>.] - ETA: 0s - loss: 0.4474 - accuracy: 0.7864

30/30 [==============================] - 2s 58ms/step - loss: 0.4472 - accuracy: 0.7869 - val_loss: 0.4602 - val_accuracy: 0.7747


Epoch 10/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4032 - accuracy: 0.8125

 3/30 [==>...........................] - ETA: 0s - loss: 0.4227 - accuracy: 0.8066

 5/30 [====>.........................] - ETA: 0s - loss: 0.4201 - accuracy: 0.8027

 7/30 [======>.......................] - ETA: 0s - loss: 0.4190 - accuracy: 0.8069

 9/30 [========>.....................] - ETA: 0s - loss: 0.4203 - accuracy: 0.8051

11/30 [==========>...................] - ETA: 0s - loss: 0.4191 - accuracy: 0.8079

13/30 [============>.................] - ETA: 0s - loss: 0.4192 - accuracy: 0.8062

15/30 [==============>...............] - ETA: 0s - loss: 0.4181 - accuracy: 0.8068

17/30 [================>.............] - ETA: 0s - loss: 0.4156 - accuracy: 0.8096

19/30 [==================>...........] - ETA: 0s - loss: 0.4160 - accuracy: 0.8086

21/30 [====================>.........] - ETA: 0s - loss: 0.4166 - accuracy: 0.8066

23/30 [======================>.......] - ETA: 0s - loss: 0.4173 - accuracy: 0.8064

25/30 [========================>.....] - ETA: 0s - loss: 0.4165 - accuracy: 0.8073

27/30 [==========================>...] - ETA: 0s - loss: 0.4155 - accuracy: 0.8088

29/30 [============================>.] - ETA: 0s - loss: 0.4142 - accuracy: 0.8111

30/30 [==============================] - 2s 58ms/step - loss: 0.4141 - accuracy: 0.8113 - val_loss: 0.4352 - val_accuracy: 0.7983


Epoch 11/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4239 - accuracy: 0.8047

 3/30 [==>...........................] - ETA: 0s - loss: 0.4131 - accuracy: 0.8118

 5/30 [====>.........................] - ETA: 0s - loss: 0.4039 - accuracy: 0.8172

 7/30 [======>.......................] - ETA: 0s - loss: 0.3946 - accuracy: 0.8245

 9/30 [========>.....................] - ETA: 0s - loss: 0.3939 - accuracy: 0.8283

11/30 [==========>...................] - ETA: 0s - loss: 0.3927 - accuracy: 0.8279

13/30 [============>.................] - ETA: 0s - loss: 0.3919 - accuracy: 0.8290

15/30 [==============>...............] - ETA: 0s - loss: 0.3887 - accuracy: 0.8299

17/30 [================>.............] - ETA: 0s - loss: 0.3878 - accuracy: 0.8289

19/30 [==================>...........] - ETA: 0s - loss: 0.3865 - accuracy: 0.8281

21/30 [====================>.........] - ETA: 0s - loss: 0.3856 - accuracy: 0.8290

23/30 [======================>.......] - ETA: 0s - loss: 0.3856 - accuracy: 0.8291

25/30 [========================>.....] - ETA: 0s - loss: 0.3846 - accuracy: 0.8302

27/30 [==========================>...] - ETA: 0s - loss: 0.3836 - accuracy: 0.8316

29/30 [============================>.] - ETA: 0s - loss: 0.3832 - accuracy: 0.8315

30/30 [==============================] - 2s 57ms/step - loss: 0.3837 - accuracy: 0.8312 - val_loss: 0.4113 - val_accuracy: 0.8074


Epoch 12/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3812 - accuracy: 0.8340

 3/30 [==>...........................] - ETA: 0s - loss: 0.3693 - accuracy: 0.8411

 5/30 [====>.........................] - ETA: 0s - loss: 0.3576 - accuracy: 0.8508

 7/30 [======>.......................] - ETA: 0s - loss: 0.3592 - accuracy: 0.8493

 9/30 [========>.....................] - ETA: 0s - loss: 0.3581 - accuracy: 0.8494

11/30 [==========>...................] - ETA: 0s - loss: 0.3595 - accuracy: 0.8457

13/30 [============>.................] - ETA: 0s - loss: 0.3608 - accuracy: 0.8454

15/30 [==============>...............] - ETA: 0s - loss: 0.3586 - accuracy: 0.8470

17/30 [================>.............] - ETA: 0s - loss: 0.3582 - accuracy: 0.8474

19/30 [==================>...........] - ETA: 0s - loss: 0.3563 - accuracy: 0.8484

21/30 [====================>.........] - ETA: 0s - loss: 0.3576 - accuracy: 0.8484

23/30 [======================>.......] - ETA: 0s - loss: 0.3569 - accuracy: 0.8483

25/30 [========================>.....] - ETA: 0s - loss: 0.3557 - accuracy: 0.8492

27/30 [==========================>...] - ETA: 0s - loss: 0.3560 - accuracy: 0.8482

29/30 [============================>.] - ETA: 0s - loss: 0.3562 - accuracy: 0.8480

30/30 [==============================] - 2s 58ms/step - loss: 0.3558 - accuracy: 0.8482 - val_loss: 0.3910 - val_accuracy: 0.8152


Epoch 13/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3506 - accuracy: 0.8438

 3/30 [==>...........................] - ETA: 0s - loss: 0.3487 - accuracy: 0.8509

 5/30 [====>.........................] - ETA: 0s - loss: 0.3413 - accuracy: 0.8539

 7/30 [======>.......................] - ETA: 0s - loss: 0.3468 - accuracy: 0.8502

 9/30 [========>.....................] - ETA: 0s - loss: 0.3471 - accuracy: 0.8500

11/30 [==========>...................] - ETA: 0s - loss: 0.3454 - accuracy: 0.8501

13/30 [============>.................] - ETA: 0s - loss: 0.3433 - accuracy: 0.8532

15/30 [==============>...............] - ETA: 0s - loss: 0.3408 - accuracy: 0.8548

17/30 [================>.............] - ETA: 0s - loss: 0.3392 - accuracy: 0.8571

19/30 [==================>...........] - ETA: 0s - loss: 0.3368 - accuracy: 0.8572

21/30 [====================>.........] - ETA: 0s - loss: 0.3348 - accuracy: 0.8572

23/30 [======================>.......] - ETA: 0s - loss: 0.3325 - accuracy: 0.8575

25/30 [========================>.....] - ETA: 0s - loss: 0.3331 - accuracy: 0.8574

27/30 [==========================>...] - ETA: 0s - loss: 0.3317 - accuracy: 0.8589

29/30 [============================>.] - ETA: 0s - loss: 0.3312 - accuracy: 0.8609

30/30 [==============================] - 2s 57ms/step - loss: 0.3305 - accuracy: 0.8611 - val_loss: 0.3727 - val_accuracy: 0.8270


Epoch 14/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3262 - accuracy: 0.8770

 3/30 [==>...........................] - ETA: 0s - loss: 0.3205 - accuracy: 0.8711

 5/30 [====>.........................] - ETA: 0s - loss: 0.3152 - accuracy: 0.8676

 7/30 [======>.......................] - ETA: 0s - loss: 0.3119 - accuracy: 0.8703

 9/30 [========>.....................] - ETA: 0s - loss: 0.3130 - accuracy: 0.8728

11/30 [==========>...................] - ETA: 0s - loss: 0.3059 - accuracy: 0.8773

13/30 [============>.................] - ETA: 0s - loss: 0.3081 - accuracy: 0.8771

15/30 [==============>...............] - ETA: 0s - loss: 0.3080 - accuracy: 0.8768

17/30 [================>.............] - ETA: 0s - loss: 0.3070 - accuracy: 0.8774

19/30 [==================>...........] - ETA: 0s - loss: 0.3076 - accuracy: 0.8764

21/30 [====================>.........] - ETA: 0s - loss: 0.3078 - accuracy: 0.8760

23/30 [======================>.......] - ETA: 0s - loss: 0.3078 - accuracy: 0.8750

25/30 [========================>.....] - ETA: 0s - loss: 0.3068 - accuracy: 0.8757

27/30 [==========================>...] - ETA: 0s - loss: 0.3080 - accuracy: 0.8743

29/30 [============================>.] - ETA: 0s - loss: 0.3073 - accuracy: 0.8743

30/30 [==============================] - 2s 58ms/step - loss: 0.3071 - accuracy: 0.8746 - val_loss: 0.3602 - val_accuracy: 0.8455


Epoch 15/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2841 - accuracy: 0.8984

 3/30 [==>...........................] - ETA: 0s - loss: 0.2973 - accuracy: 0.8926

 5/30 [====>.........................] - ETA: 0s - loss: 0.2936 - accuracy: 0.8883

 7/30 [======>.......................] - ETA: 0s - loss: 0.2857 - accuracy: 0.8923

 9/30 [========>.....................] - ETA: 0s - loss: 0.2871 - accuracy: 0.8891

11/30 [==========>...................] - ETA: 0s - loss: 0.2872 - accuracy: 0.8864

13/30 [============>.................] - ETA: 0s - loss: 0.2859 - accuracy: 0.8851

15/30 [==============>...............] - ETA: 0s - loss: 0.2858 - accuracy: 0.8865

17/30 [================>.............] - ETA: 0s - loss: 0.2860 - accuracy: 0.8847

19/30 [==================>...........] - ETA: 0s - loss: 0.2884 - accuracy: 0.8834

21/30 [====================>.........] - ETA: 0s - loss: 0.2881 - accuracy: 0.8843

23/30 [======================>.......] - ETA: 0s - loss: 0.2883 - accuracy: 0.8830

25/30 [========================>.....] - ETA: 0s - loss: 0.2873 - accuracy: 0.8841

27/30 [==========================>...] - ETA: 0s - loss: 0.2867 - accuracy: 0.8840

29/30 [============================>.] - ETA: 0s - loss: 0.2870 - accuracy: 0.8838

30/30 [==============================] - 2s 58ms/step - loss: 0.2872 - accuracy: 0.8840 - val_loss: 0.3445 - val_accuracy: 0.8462


Epoch 16/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2690 - accuracy: 0.8945

 3/30 [==>...........................] - ETA: 0s - loss: 0.2714 - accuracy: 0.8919

 5/30 [====>.........................] - ETA: 0s - loss: 0.2743 - accuracy: 0.8902

 7/30 [======>.......................] - ETA: 0s - loss: 0.2797 - accuracy: 0.8867

 9/30 [========>.....................] - ETA: 0s - loss: 0.2751 - accuracy: 0.8882

11/30 [==========>...................] - ETA: 0s - loss: 0.2722 - accuracy: 0.8908

13/30 [============>.................] - ETA: 0s - loss: 0.2709 - accuracy: 0.8932

15/30 [==============>...............] - ETA: 0s - loss: 0.2697 - accuracy: 0.8926

17/30 [================>.............] - ETA: 0s - loss: 0.2686 - accuracy: 0.8928

19/30 [==================>...........] - ETA: 0s - loss: 0.2704 - accuracy: 0.8916

21/30 [====================>.........] - ETA: 0s - loss: 0.2697 - accuracy: 0.8911

23/30 [======================>.......] - ETA: 0s - loss: 0.2691 - accuracy: 0.8923

25/30 [========================>.....] - ETA: 0s - loss: 0.2695 - accuracy: 0.8920

27/30 [==========================>...] - ETA: 0s - loss: 0.2682 - accuracy: 0.8935

29/30 [============================>.] - ETA: 0s - loss: 0.2683 - accuracy: 0.8939

30/30 [==============================] - 2s 58ms/step - loss: 0.2678 - accuracy: 0.8942 - val_loss: 0.3333 - val_accuracy: 0.8538


Epoch 17/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2602 - accuracy: 0.8867

 3/30 [==>...........................] - ETA: 0s - loss: 0.2483 - accuracy: 0.9030

 5/30 [====>.........................] - ETA: 0s - loss: 0.2461 - accuracy: 0.9027

 7/30 [======>.......................] - ETA: 0s - loss: 0.2481 - accuracy: 0.9018

 9/30 [========>.....................] - ETA: 0s - loss: 0.2498 - accuracy: 0.9019

11/30 [==========>...................] - ETA: 0s - loss: 0.2494 - accuracy: 0.9027

13/30 [============>.................] - ETA: 0s - loss: 0.2506 - accuracy: 0.9016

15/30 [==============>...............] - ETA: 0s - loss: 0.2498 - accuracy: 0.9018

17/30 [================>.............] - ETA: 0s - loss: 0.2509 - accuracy: 0.9020

19/30 [==================>...........] - ETA: 0s - loss: 0.2506 - accuracy: 0.9020

21/30 [====================>.........] - ETA: 0s - loss: 0.2527 - accuracy: 0.9001

23/30 [======================>.......] - ETA: 0s - loss: 0.2502 - accuracy: 0.9013

25/30 [========================>.....] - ETA: 0s - loss: 0.2514 - accuracy: 0.9009

27/30 [==========================>...] - ETA: 0s - loss: 0.2506 - accuracy: 0.9009

29/30 [============================>.] - ETA: 0s - loss: 0.2507 - accuracy: 0.9009

30/30 [==============================] - 2s 58ms/step - loss: 0.2505 - accuracy: 0.9010 - val_loss: 0.3243 - val_accuracy: 0.8557


Epoch 18/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2177 - accuracy: 0.9219

 3/30 [==>...........................] - ETA: 0s - loss: 0.2405 - accuracy: 0.9108

 5/30 [====>.........................] - ETA: 0s - loss: 0.2416 - accuracy: 0.9059

 7/30 [======>.......................] - ETA: 0s - loss: 0.2401 - accuracy: 0.9074

 9/30 [========>.....................] - ETA: 0s - loss: 0.2369 - accuracy: 0.9067

11/30 [==========>...................] - ETA: 0s - loss: 0.2399 - accuracy: 0.9050

13/30 [============>.................] - ETA: 0s - loss: 0.2374 - accuracy: 0.9062

15/30 [==============>...............] - ETA: 0s - loss: 0.2358 - accuracy: 0.9076

17/30 [================>.............] - ETA: 0s - loss: 0.2354 - accuracy: 0.9073

19/30 [==================>...........] - ETA: 0s - loss: 0.2375 - accuracy: 0.9062

21/30 [====================>.........] - ETA: 0s - loss: 0.2373 - accuracy: 0.9068

23/30 [======================>.......] - ETA: 0s - loss: 0.2376 - accuracy: 0.9074

25/30 [========================>.....] - ETA: 0s - loss: 0.2356 - accuracy: 0.9084

27/30 [==========================>...] - ETA: 0s - loss: 0.2348 - accuracy: 0.9079

29/30 [============================>.] - ETA: 0s - loss: 0.2352 - accuracy: 0.9072

30/30 [==============================] - 2s 57ms/step - loss: 0.2351 - accuracy: 0.9073 - val_loss: 0.3172 - val_accuracy: 0.8634


Epoch 19/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2307 - accuracy: 0.9141

 3/30 [==>...........................] - ETA: 0s - loss: 0.2258 - accuracy: 0.9173

 5/30 [====>.........................] - ETA: 0s - loss: 0.2269 - accuracy: 0.9180

 7/30 [======>.......................] - ETA: 0s - loss: 0.2300 - accuracy: 0.9155

 9/30 [========>.....................] - ETA: 0s - loss: 0.2297 - accuracy: 0.9167

11/30 [==========>...................] - ETA: 0s - loss: 0.2300 - accuracy: 0.9135

13/30 [============>.................] - ETA: 0s - loss: 0.2270 - accuracy: 0.9153

15/30 [==============>...............] - ETA: 0s - loss: 0.2236 - accuracy: 0.9160

17/30 [================>.............] - ETA: 0s - loss: 0.2212 - accuracy: 0.9168

19/30 [==================>...........] - ETA: 0s - loss: 0.2225 - accuracy: 0.9146

21/30 [====================>.........] - ETA: 0s - loss: 0.2208 - accuracy: 0.9147

23/30 [======================>.......] - ETA: 0s - loss: 0.2197 - accuracy: 0.9156

25/30 [========================>.....] - ETA: 0s - loss: 0.2189 - accuracy: 0.9163

27/30 [==========================>...] - ETA: 0s - loss: 0.2194 - accuracy: 0.9162

29/30 [============================>.] - ETA: 0s - loss: 0.2206 - accuracy: 0.9156

30/30 [==============================] - 2s 58ms/step - loss: 0.2209 - accuracy: 0.9154 - val_loss: 0.3108 - val_accuracy: 0.8660


Epoch 20/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1873 - accuracy: 0.9316

 3/30 [==>...........................] - ETA: 0s - loss: 0.1981 - accuracy: 0.9251

 5/30 [====>.........................] - ETA: 0s - loss: 0.2081 - accuracy: 0.9203

 7/30 [======>.......................] - ETA: 0s - loss: 0.2067 - accuracy: 0.9210

 9/30 [========>.....................] - ETA: 0s - loss: 0.2073 - accuracy: 0.9201

11/30 [==========>...................] - ETA: 0s - loss: 0.2096 - accuracy: 0.9208

13/30 [============>.................] - ETA: 0s - loss: 0.2089 - accuracy: 0.9229

15/30 [==============>...............] - ETA: 0s - loss: 0.2108 - accuracy: 0.9212

17/30 [================>.............] - ETA: 0s - loss: 0.2123 - accuracy: 0.9200

19/30 [==================>...........] - ETA: 0s - loss: 0.2111 - accuracy: 0.9201

21/30 [====================>.........] - ETA: 0s - loss: 0.2105 - accuracy: 0.9201

23/30 [======================>.......] - ETA: 0s - loss: 0.2105 - accuracy: 0.9203

25/30 [========================>.....] - ETA: 0s - loss: 0.2097 - accuracy: 0.9205

27/30 [==========================>...] - ETA: 0s - loss: 0.2086 - accuracy: 0.9219

29/30 [============================>.] - ETA: 0s - loss: 0.2084 - accuracy: 0.9225

30/30 [==============================] - 2s 57ms/step - loss: 0.2082 - accuracy: 0.9224 - val_loss: 0.3058 - val_accuracy: 0.8676


## 评估模型

我们来看下模型的表现如何。将返回两个值。损失值（loss）（一个表示误差的数字，值越低越好）与准确率（accuracy）。

In [11]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3208 - accuracy: 0.8546


loss: 0.321
accuracy: 0.855


这种十分朴素的方法得到了约 87% 的准确率（accuracy）。若采用更好的方法，模型的准确率应当接近 95%。

## 进一步阅读

有关使用字符串输入的更一般方法，以及对训练期间准确率（accuracy）和损失值（loss）更详细的分析，请参阅[此处](https://tensorflow.google.cn/tutorials/keras/basic_text_classification)。